File: medium/_2181.py
# Definition for singly-linked list.
class _2181Node(object):
    def __init__(self, val=None, next=None):
        self.val = val
        self.next = next

class Solution:

    def __init__(self):
      self.head = None
    # helper function to insert nodes
    def insert(self, val):
        if (self.head == None): 
            self.head = _2181Node(val)
            return
        else:
            curr = self.head
            while(curr.next):
                curr = curr.next
            curr.next = _2181Node(val)
    
    def mergeNodes(self, head):

        res = _2181Node() # head of result linkedlist
        tail = _2181Node() # helper pointer to insert to res
        headTrav = head # to traverse given linkedlist
        sum = 0 # to count values of nodes between zeros

        while(headTrav.next != None):
            if (headTrav.next.val == 0):
                sum+=headTrav.val
                curr = _2181Node(sum)
                # insert to result linkedlist
                if (res.val == None):
                    res = curr
                    tail = curr
                
                else:
                    tail.next = curr
                    tail = curr; 
                
                sum=0 # reset when reach zero node
                
            else:
                # sum values of nodes until zero node
                sum+= headTrav.val
            
            headTrav = headTrav.next

        return res

ll = Solution()
ll.insert(0)
ll.insert(3)
ll.insert(1)
ll.insert(0)
ll.insert(4)
ll.insert(5)
ll.insert(2)
ll.insert(0)

res = ll.mergeNodes(ll.head)

while (res != None):
    print(res.val)
    res = res.next 

'''
You are given the head of a linked list, which contains a series of integers separated by 0's. 
The beginning and end of the linked list will have Node.val == 0.

For every two consecutive 0's, merge all the nodes lying in between them into a single node whose value is the sum of all the merged nodes. 
The modified list should not contain any 0's.

Return the head of the modified linked list.
''' 

File: medium/_2181.cpp
#include <iostream>

using namespace std;

// Definition for singly-linked list.
struct _2181Node {
    int val;
    _2181Node *next;
    _2181Node() : val(0), next(nullptr) {}
    _2181Node(int x) : val(x), next(nullptr) {}
    _2181Node(int x, _2181Node *next) : val(x), next(next) {}
};

struct LinkedList {
    public:
        _2181Node *head, *tail;
    public:
        LinkedList() {
        head = NULL;
        tail = NULL;
        }
                // helper function to insert node for testing
        void insert(int val) {
            _2181Node *newNode = new _2181Node(val);

            if (head == NULL) {
                head = newNode;
                tail = newNode;
            }
            else {
                tail->next = newNode;
                tail = newNode;
            }
        }

        _2181Node *mergeNodes(_2181Node *head) {
            _2181Node *traverse = head->next;
            _2181Node *res = NULL;
            _2181Node *resTraverse;
            int sum = 0;
            while (traverse != NULL) {
                if (traverse->val != 0) {
                    sum+=traverse->val;
                } else if (res == NULL) {
                    res = new _2181Node(sum);
                    resTraverse = res;
                    sum=0;
                } else {
                    resTraverse->next = new _2181Node(sum);
                    resTraverse = resTraverse->next;
                    sum=0;
                }

                traverse = traverse->next;
            }
            return res;
        }
};


int main() {
    LinkedList nums;
    
    // populate the list
    nums.insert(0);
    nums.insert(1);
    nums.insert(0);
    nums.insert(3);
    nums.insert(0);
    nums.insert(2);
    nums.insert(2);
    nums.insert(0);

    _2181Node *res = nums.mergeNodes(nums.head);
    // display nodes without mid
    while (res != NULL) {
        cout << res->val << endl;
        res = res->next;
    }

    return 0;
}

// You are given the head of a linked list, which contains a series of integers separated by 0's. The beginning and end of the linked list will have Node.val == 0.

// For every two consecutive 0's, merge all the nodes lying in between them into a single node whose value is the sum of all the merged nodes. The modified list should not contain any 0's.

// Return the head of the modified linked list.
